File: leetcode/0131-palindrome-partitioning/solution.cpp
#include <string>
#include <vector>
using namespace std;

class Solution {
public:
    vector<vector<string>> partition(string s) {
        vector<vector<string>> result;
        vector<string> path;
        backtrack(s, 0, path, result);
        return result;
    }

private:
    void backtrack(const string& s, int start, vector<string>& path, vector<vector<string>>& result) {
        int n = (int)s.size();
        if (start == n) {
            result.push_back(path);
            return;
        }
        for (int end = start + 1; end <= n; end++) {
            if (isPalindrome(s, start, end - 1)) {
                path.push_back(s.substr(start, end - start));
                backtrack(s, end, path, result);
                path.pop_back();
            }
        }
    }

    bool isPalindrome(const string& s, int left, int right) {
        while (left < right) {
            if (s[left] != s[right]) {
                return false;
            }
            left++;
            right--;
        }
        return true;
    }
};

File: leetcode/0131-palindrome-partitioning/solution.py
from typing import List


class Solution:
    def partition(self, s: str) -> List[List[str]]:
        result = []
        path = []
        n = len(s)

        def is_palindrome(left: int, right: int) -> bool:
            while left < right:
                if s[left] != s[right]:
                    return False
                left += 1
                right -= 1
            return True

        def backtrack(start: int) -> None:
            if start == n:
                result.append(path[:])
                return
            for end in range(start + 1, n + 1):
                if is_palindrome(start, end - 1):
                    path.append(s[start:end])
                    backtrack(end)
                    path.pop()

        backtrack(0)
        return result
